XAES-256-GCM.md

```diff
@@ -0,0 +1,45 @@
+This document specifies the XAES-256-GCM authenticated encryption with additional data algorithm, based on the composition of a standard NIST SP 800-108r1 KDF and the standard NIST AES-256-GCM AEAD.
+
+The XAES-256-GCM inputs are a 256-bit key, a 192-bit nonce, a plaintext of up to approximately 64GiB, and additional data of up to 2 EiB.
+
+Unlike AES-256-GCM, the XAES-256-GCM nonce can be randomly generated for a virtually unlimited number of messages (2⁸⁰ messages with collision risk 2⁻³²). Only a 256-bit key version is specified, which provides a comfortable multi-user security margin. Like AES-256-GCM, XAES-256-GCM is not nonce misuse-resistant, nor is it key-committing. Compared to AES-256-GCM, XAES-256-GCM requires three extra invocations of the AES-256 function, but has otherwise the same performance profile.
+
+## Overview
+
+XAES-256-GCM derives a subkey for use with AES-256-GCM from the input key and half the input nonce using a NIST SP 800-108r1 KDF.
+
+The counter-based KDF (NIST SP 800-108r1, Section 4.1) is instantiated with CMAC-AES256 (NIST SP 800-38B) and the input key as *Kin*, the ASCII letter `X` (0x58) as *Label*, and the first 96 bits of the input nonce as *Context* (as recommended by NIST SP 800-38B, Section 4, point 4), to produce a 256-bit derived key.
+
+Note that with a counter (*i*) size of 16 bits and omitting the optional *L* field, the AES-CMAC input totals 128 bits, which fits into a single block, mitigating the key control security issue described in NIST SP 800-38B, Section 6.7. It would in theory be possible to shrink *i* to 8 bits to fit a 8 bits *L*, but some implementations unfortunately fix the *L* size to 32 bits.
+
+The derived key and the last 96 bits of the input nonce are used to encrypt the message with AES-256-GCM.
+
+## Detailed key derivation algorithm
+
+Inputs:
+
+* 256-bit key *K*
+* 192-bit nonce *N*
+
+Outputs:
+
+* 256-bit key *Kₓ*
+* 96-bit nonce *Nₓ*
+
+Algorithm:
+
+1. *L* = AES-256ₖ(0¹²⁸)
+2. If MSB₁(*L*) = 0, then *K1* = *L* << 1;
+   Else *K1* = (*L* << 1) ⊕ 0¹²⁰10000111
+3. *M1* = 0x00 || 0x01 || `X` || 0x00 || *N*[:12]
+4. *M2* = 0x00 || 0x02 || `X` || 0x00 || *N*[:12]
+5. *Kₓ* = AES-256ₖ(*M1* ⊕ *K1*) || AES-256ₖ(*M2* ⊕ *K1*)
+6. *Nₓ* = *N*[12:]
+
+Steps 1 and 2 reproduce the CMAC subkey generation specified in NIST SP 800-38B, Section 6.1. Note that only *K1* is needed.
+
+Steps 2 and 3 compose the PRF inputs for *i* = 1, 2 according to NIST SP 800-108r1, Section 4.1.
```

Comment:
```suggestion
Steps 3 and 4 compose the PRF inputs for *i* = 1, 2 according to [NIST SP 800-108r1, Section 4.1].
```

[NIST SP 800-108r1, Section 4.1](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-108r1.pdf#%5B%7B%22num%22%3A79%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C70%2C300%2C0%5D) specifies the PRF inputs with a 2-byte suffix committing to the output length in bits. In this case the "output" here has constant length, so the length commitment is unnecessary (and dropping it lets the input fit into an AES-256 block); it might be worth mentioning this here.

Reply:
That's the *L* field mentioned in the third Overview paragraph, now expanded.